vision_evaluation/evaluators.py

```diff
@@ -1044,5 +1175,73 @@ def add_predictions(self, predictions, targets):
         for pred_mask, gt_mask in zip(predictions, targets):
             pred_mask = np.asarray(pred_mask)
             gt_mask = np.asarray(gt_mask)
-            mean_l1 = np.abs(pred_mask.astype(np.float)-gt_mask.astype(np.float)).mean()
+            mean_l1 = np.abs(
+                pred_mask.astype(np.float) - gt_mask.astype(np.float)).mean()
             self._metric_sum += mean_l1
+
+
+class GroupWiseEvaluator(Evaluator):
+    """
+    Group Wise Evaluator
+    """
+    def __init__(self, evaluator, task_name, group_name,
+                 multitask_manifest_dataset, target_task_predictions):
+        """
+        Args:
+            evaluator (vision_evaluation.Evaluator): Evaluation class object. E.g. TopAccuracyEvaluator()
+            
+            manifest_dataset (ManifestDataset): Iterable multitask ManifestDataset object.
+            
+            target_task_predictions (N, num_class):  Model output numpy array for the target task.
+            
+            task_name (Str):  Target attribute used for the classification task. E.g. "gender"
+            
+            group_name (Str): Attribute for slicing samples. E.g. "ethnicity"     
+        """
+
+        self.task_name = task_name
+        self.group_name = group_name
+        self.manifest_dataset = multitask_manifest_dataset
+        self.target_task_predictions = target_task_predictions
+        self.evaluator = evaluator
+
+    def group_by(self):
+        """Group ground truth and predictions by task and group names. 
+        """
+
+        # Step 0: Sanity Checks
+        assert len(self.manifest_dataset) == len(self.target_task_predictions)
+        assert type(self.task_name) == str
+        assert type(self.group_name) == str
+
+        # Step1: Sample datapoints by target_task and groups and save in store dict
+        self.store = collections.defaultdict(
+            lambda: collections.defaultdict(list))
+
+        for _, targets, sample_idx in self.manifest_dataset:
+
+            cls = targets[self.task_name][0]
+            grp = targets[self.group_name][0]
+
+            self.store[(cls, grp)]['idx'].append(sample_idx)
+            self.store[(cls, grp)]['targets'].append(cls)
+            self.store[(cls, grp)]['predictions'].append(
+                self.target_task_predictions[int(sample_idx)])
+
+    def add_predictions(self, predictions, targets):
+        self.evaluator.add_predictions(predictions, targets)
```

Comment:
it adds prediction results to the self.evaluator, but how it will be used? It seems it will be reset at line 1242.

Reply:
i am resetting the evaluator everytime i add prediction results for each target_task and group combination.